src/librustc/ty/instance.rs

```diff
@@ -55,9 +61,19 @@ pub enum InstanceDef<'tcx> {
     },
 
     /// `drop_in_place::<T>; None` for empty drop glue.
+    ///
+    /// NB: the type must be monomorphic for MIR shims to work.
+    // FIXME(eddyb) support generating shims for a "shallow type",
+    // e.g. `Foo<_>` or `[_]` instead of requiring a fully monomorphic
+    // `Foo<Bar>` or `[String]` etc.
     DropGlue(DefId, Option<Ty<'tcx>>),
 
     ///`<T as Clone>::clone` shim.
+    ///
+    /// NB: the type must be monomorphic for MIR shims to work.
```

Comment:
I feel like this comment isn't really explaining what's going on.  I'm trying to think how to write it and realizing that I can't quite.

Reply:
In trans, obviously, this would be monomorphic all the time. So the problem arises outside of trans, such as in the inliner.

Reply:
I guess I think the comment could at minimum be written as:

```rust
/// NB: At present, the type will always be monomorpic (we return `None` rather than resolve to a generic shim. 
```

Reply:
I don't think we should treat monomorphic codegen as "a fact of life" in rustc, since it easily leads to soundness holes (`Instance::resolve` until recently was wrong about specialization, for example).

Instead, we should assume that everything is as absurdly polymorphic as possible, and restrict it only when it's unworkable.

You're right that the comments on the variants are useless though.

Note that the comments are intended for *creators* of the variants, *not* consumers. Anyone can theoretically make an `Instance` out of thin air, and I need to figure out a good way to document the fact that generic parameters are not supported.

And they're not supported because... they would have to be substitutable away by the `Substs` in `Instance`, which are for the definition of the shimmed thing (e.g. `clone` method of trait `Clone`, so the only available parameter would be `Self`, which ironically is the type itself).

But the symptom would be that the MIR body of the shim would be malformed, since it would likely refer to type parameters that don't even fit the `Substs`.

Overall this is annoying to explain and I just wish it was easier to implement the "correct thing" where the shims are *never* monomorphic (except for user ADTs with no generics).